### PR TITLE
[JENKINS-25326] Elevates to SYSTEM when handling throttling for builds

### PR DESCRIPTION
**Pull request by ikedam**
[JENKINS-25326](https://issues.jenkins-ci.org/browse/JENKINS-25326)

Throttling doesn't work for projects in cloudbees-folder with latest Jenkins.
As far as I tested, it doesn't work since Jenkins 1.536:

| Jenlkins | Throttling in folder | Note |
|:---------|:---------------------|:-----|
| 1.532    | Works                | |
| 1.535    | Works                | |
| 1.536    | **Doesn't work**     | |
| 1.554    | **Doesn't work**     | |
| 1.642    | **Doesn't work**     | Requires some changes in test codes to compile |

I fix that problem in followng steps in this request:

1. Add a commit to add a test to reproduce the problem. This requires to change the targeted Jenkins version to >= 1.536.
2. Add a commit to fix the problem.
3. Add a commit to revert changes to reproduce the problem, including the change of targeted Jenkins version.

### CODE REVIEW

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java [hunk 1]
```diff
@@ -80,8 +84,19 @@ private void setupSlave() throws Exception {
...
+    /**
+     * setup security so that no one except SYSTEM has any permissions.
+     * should be called after {@link #setupSlave()}
+     */
+    private void setupSecurity() {
+        jenkins.setSecurityRealm(createDummySecurityRealm());
+        GlobalMatrixAuthorizationStrategy auth = new GlobalMatrixAuthorizationStrategy();
+        jenkins.setAuthorizationStrategy(auth);
+    }
+
```
Comment (ikedam) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java]: The problem is a permission issue, and requires no READ permission for `ANONYMOUS` to reproduce.

Creating a slave triggers serializing Jenkins configuration and results an exception as `DummySecurityRealm` is not capable to be serialized.
So you have to call `setupSecurity` after `setupSlave`.

**Comment by ikedam** [comment]
d2f6ca409155c32dcb79d32c835614b95d9afc38 failed as throttle-concurrent couldn't access projects in cloudbees-folders. Throttle didn't work and builds ran at the same time
https://jenkins.ci.cloudbees.com/job/plugins/job/throttle-concurrent-builds-plugin/org.jenkins-ci.plugins$throttle-concurrents/83/testReport/junit/hudson.plugins.throttleconcurrents/ThrottleIntegrationTest/testThrottlingWithCategoryInFolder/
```
junit.framework.AssertionFailedError: expected:<1> but was:<2>
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.failNotEquals(Assert.java:329)
	at junit.framework.Assert.assertEquals(Assert.java:78)
	at junit.framework.Assert.assertEquals(Assert.java:234)
	at junit.framework.Assert.assertEquals(Assert.java:241)
	at junit.framework.TestCase.assertEquals(TestCase.java:409)
	at hudson.plugins.throttleconcurrents.ThrottleIntegrationTest.testThrottlingWithCategoryInFolder(ThrottleIntegrationTest.java:216)
```

**Comment by ikedam** [comment]
09234147a9bf23de0cd40cfa420cb8d4e97a4a6e fixed the problem.

**Comment by ikedam** [comment]
Reverted changes necessary to reproduce the problem but unnecessary for the fix itself.
The request is ready to review.

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java [hunk 2]
```diff
@@ -21,11 +23,34 @@
...
 @Extension
 public class ThrottleQueueTaskDispatcher extends QueueTaskDispatcher {
 
     @Override
     public CauseOfBlockage canTake(Node node, Task task) {
+        if (Jenkins.getAuthentication() == ACL.SYSTEM) {
+            return canTakeImpl(node, task);
+        }
+        
+        // Throttle-concurrent-builds requires READ permissions for all projects.
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: Should work for old cores

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java [hunk 3]
```diff
@@ -37,9 +38,14 @@
 import java.util.Arrays;
 import java.util.Collections;
 
+import org.jvnet.hudson.test.Bug;
 import org.jvnet.hudson.test.HudsonTestCase;
 import org.jvnet.hudson.test.SleepBuilder;
 
+/*
+import com.cloudbees.hudson.plugins.folder.Folder;
```
Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java]: dead code

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcherTest.java [hunk 4]
```diff
@@ -358,7 +359,13 @@ private String configureLogger()
                     input.setValueAttribute(logger);
                 }
                 HtmlSelect select = form.getSelectByName("level");
-                HtmlOption option = select.getOptionByValue("fine");
+                HtmlOption option;
```
Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcherTest.java]: Uh ok. new HTML Unit will likely break the stuff again

File: src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java (hunk 2 above)
Comment (ikedam) [inline src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java]: I cannot get what you want to say.
Do you mean this causes functionality problems with 1.424?
I don't think so. Actually unit tests passed even with this change.
Please let me know problems you worry about.

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcherTest.java (hunk 4 above)
Comment (ikedam) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcherTest.java]: Should I do something for this?

Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcherTest.java]: No, this is an AI for a person, who will be performing the upgrade

**Comment by oleg-nenashev** [comment]
Tested the PR manually. :+1: , it's ready to go if the merge conflict gets fixed

File: src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java [hunk 5]
```diff
@@ -147,4 +165,59 @@ public void testThrottlingWithCategory() throws Exception {
         // throttled, and only one build runs at the same time.
         assertEquals(1, waterMark.getExecutorWaterMark());
     }
+    
+    /*
+    // Requires Jenkins >= 1.480.3 and cloudbees-folder-plugin
```
Comment (oleg-nenashev) [inline src/test/java/hudson/plugins/throttleconcurrents/ThrottleIntegrationTest.java]: Nice2Have: TODO marker

**Comment by oleg-nenashev** [comment]
I'm integrating this PR right now

**Comment by oleg-nenashev** [comment]
Merged via #41 
Thanks for the fix!

**Pull request merged**